### PR TITLE
Use more precise span when checking type definitions

### PR DESCRIPTION
**Pull request by sanxiyn**
When checking field types of struct or enum definitions, use spans for field types instead of the entire item.

### CODE REVIEW

File: src/librustc_typeck/check/wf.rs [hunk 1]
```diff
@@ -483,11 +480,10 @@ pub struct BoundsChecker<'cx,'tcx:'cx> {
 
 impl<'cx,'tcx> BoundsChecker<'cx,'tcx> {
     pub fn new(fcx: &'cx FnCtxt<'cx,'tcx>,
-               span: Span,
                scope: ast::NodeId,
                cache: Option<&'cx mut HashSet<Ty<'tcx>>>)
                -> BoundsChecker<'cx,'tcx> {
-        BoundsChecker { fcx: fcx, span: span, scope: scope,
+        BoundsChecker { fcx: fcx, span: DUMMY_SP, scope: scope,
```
Comment (alexcrichton) [inline src/librustc_typeck/check/wf.rs]: If this `span` field is never used, can it be deleted? If it is used, it seems odd to use `DUMMY_SP` here?

**Comment by arielb1** [comment]
Needs a rebase.

**Comment by arielb1** [comment]
Actually, I would also want tests that this feature works on both structs and enums, both tuple-like and dictionary-like.

File: src/librustc_typeck/check/wf.rs (hunk 1 above)
Comment (sanxiyn) [inline src/librustc_typeck/check/wf.rs]: After the patch, it is used, but not initialized here.

**Comment by sanxiyn** [comment]
Rebased and added tests.

**Pull request merged**